org.spoofax.terms/src/org/spoofax/terms/StrategoArrayList.java

```diff
@@ -1,289 +1,96 @@
-        this(terms, annotations, 0, terms.length, new ITermAttachment[terms.length]);
+    public StrategoArrayList(IStrategoTerm... backingArray) {
+        this(backingArray, null);
     }
 
-    protected StrategoArrayList(IStrategoTerm[] terms, IStrategoList annotations, int offset, int endOffset) {
-        this(terms, annotations, offset, endOffset, new ITermAttachment[terms.length]);
+    public StrategoArrayList(IStrategoTerm[] backingArray, IStrategoList annotations) {
+        this(backingArray, annotations, 0, backingArray.length);
     }
 
-    protected StrategoArrayList(IStrategoTerm[] terms, IStrategoList annotations, int offset, int endOffset, ITermAttachment[] tailAttachments) {
+    protected StrategoArrayList(IStrategoTerm[] backingArray, IStrategoList annotations, int offset, int endOffset) {
         super(annotations);
-        if(offset > terms.length) {
+        if(offset > backingArray.length) {
             throw new IllegalArgumentException(
-                "Offset (" + offset + ") is larger than backing array (" + terms.length + ")");
+                "Offset (" + offset + ") is larger than backing array (" + backingArray.length + ")");
         }
-        if(endOffset > terms.length) {
+        if(endOffset > backingArray.length) {
             throw new IllegalArgumentException(
-                "Length (" + endOffset + ") is larger than backing array (" + terms.length + ")");
+                "Length (" + endOffset + ") is larger than backing array (" + backingArray.length + ")");
         }
-        this.terms = terms;
+        this.backingArray = backingArray;
         this.offset = offset;
         this.endOffset = endOffset;
-        this.tailAttachments = tailAttachments;
-        if(offset > 0) {
-            final ITermAttachment attachment = tailAttachments[offset-1];
-            if(attachment != null) {
-                super.putAttachment(attachment);
-            }
-        }
-    }
-
-    public static StrategoArrayList fromCollection(Collection<? extends IStrategoTerm> terms) {
-        return new StrategoArrayList(terms.toArray(EMPTY_TERM_ARRAY));
-    }
-
-    public static StrategoArrayListBuilder arrayListBuilder() {
-        return new StrategoArrayListBuilder(16);
-    }
-
-    public static StrategoArrayListBuilder arrayListBuilder(int size) {
-        return new StrategoArrayListBuilder(size);
-    }
-
-    @Override public int getSubtermCount() {
-        return endOffset - offset;
-    }
-
-    @Override public IStrategoTerm getSubterm(int index) {
-        if(index < getSubtermCount()) {
-            return terms[offset + index];
-        } else {
-            throw new IndexOutOfBoundsException();
-        }
+        this.tailAttachments = null;
     }
 
     @Override
-    public IStrategoTerm[] getAllSubterms() {
-        return Arrays.copyOfRange(terms, offset, endOffset);
+    public IStrategoTerm[] internalGetBackingArray() {
+        return backingArray;
```

Comment:
Do note that this is semantically different: I can change the elements in the list by changing the elements in the returned array, whereas I could not before. Of course, this is a performance trade-off. Therefore, perhaps add a comment with the _why_.

Reply:
`internalGetBackingArray` is not a replacement for `getAllSubterms`, they just ended up side-by-side in the diff. `getAllSubterms` was just moved the shared interface. I can add more javadocs to discourage usage of the `internal` methods.